### PR TITLE
Fix Index is Greater than the items available from a VirtualizePanelAdaptater for WASM

### PR DESCRIPTION
**Pull request by GuillaumeSE**
GitHub Issue (If applicable): #

https://github.com/unoplatform/private/issues/60
## PR Type

What kind of change does this PR introduce?

## What is the current behavior?

In WASM the index is superior to the VirtualizePanelAdaptater Item causing a OutofRangeException.
https://gist.github.com/ghuntley/90eb1a9af59c5be87a43df33323e059b

## What is the new behavior?
Verify if the Index is not bigger than the list's length

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../README.md#supported)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by GuillaumeSE** [comment]
This may have issue on the other platforms as i make return a null value if the index is not in the range.
Please advise :)

File: src/SamplesApp/UITests.Shared/UITests.Shared.projitems [hunk 1]
```diff
@@ -4451,6 +4456,12 @@
     <Compile Update="$(MSBuildThisFileDirectory)Windows_UI_Xaml_Controls\ToolTip\TextOnlyToolTipSample.xaml.cs">
       <DependentUpon>TextOnlyToolTipSample.xaml</DependentUpon>
     </Compile>
+    <Compile Update="C:\P\Uno\Uno.UI\src\SamplesApp\UITests.Shared\Windows_UI_Xaml_Controls\ListView\ListView_VirtualizePanelAdaptaterIdCache - Copy.xaml.cs">
```
[the remark this answers is not in the document]
Comment (GuillaumeSE) [reply src/SamplesApp/UITests.Shared/UITests.Shared.projitems]: Alright thanks :)

File: src/SamplesApp/UITests.Shared/UITests.Shared.projitems [hunk 2]
```diff
@@ -4461,6 +4466,12 @@
     <Compile Update="$(MSBuildThisFileDirectory)Windows_UI_Xaml_Controls\ToolTip\TextOnlyToolTipSample.xaml.cs">
       <DependentUpon>TextOnlyToolTipSample.xaml</DependentUpon>
     </Compile>
+    <Compile Update="C:\P\Uno\Uno.UI\src\SamplesApp\UITests.Shared\Windows_UI_Xaml_Controls\ListView\ListView_VirtualizePanelAdaptaterIdCache.xaml.cs">
```
Comment (davidjohnoliver) [inline src/SamplesApp/UITests.Shared/UITests.Shared.projitems]: should not be an absolute path (`C:\etc`) here - this is coming from a VS bug, you can fix it manually

File: src/SamplesApp/UITests.Shared/UITests.Shared.projitems [hunk 3]
```diff
@@ -4461,6 +4466,12 @@
...
+      <DependentUpon>ListView_VirtualizePanelAdaptaterIdCache.xaml</DependentUpon>
+    </Compile>
+    <Compile Update="C:\src\Uno2\src\SamplesApp\UITests.Shared\Windows_UI_Xaml_Controls\ComboBox\ComboBox_ItemDataContext.xaml.cs">
```
Comment (davidjohnoliver) [inline src/SamplesApp/UITests.Shared/UITests.Shared.projitems]: ditto

File: src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelAdapter.wasm.cs [hunk 4]
```diff
@@ -15,8 +15,8 @@ internal class VirtualizingPanelGenerator
 		private const int CacheLimit = 10;
 		private const int NoTemplateItemId = -1;
 		private readonly VirtualizingPanelLayout _owner;
-
 		private readonly Dictionary<int, Stack<FrameworkElement>> _itemContainerCache = new Dictionary<int, Stack<FrameworkElement>>();
+		private Dictionary<int, int> _idCache = new Dictionary<int, int>();
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelAdapter.wasm.cs]:
```suggestion
		/// <summary>
		/// Caching the id is more efficient, and also important in the case of the ItemsSource changing, when the (former) item may no longer be in the new collection.
		/// </summary>
		private Dictionary<int, int> _idCache = new Dictionary<int, int>();
```

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ListViewTests/UnoSamples_Tests.ListView.cs [hunk 5]
```diff
@@ -70,5 +70,17 @@ void TabWaitAndThenScreenshot(string buttonName)
...
+		[Test]
+		public void ListView_VirtualizePanelAdaptaterCache()
+		{
+			Run("UITests.Shared.Windows_UI_Xaml_Controls.ListView.ListView_VirtualizePanelAdaptaterCache");
+
+			_app.WaitForElement(_app.Marked("MyListView"));
+			var theListView = _app.Marked("MyListView");
+			_app.WaitForElement(_app.Marked("TextResult"));
+			var theTextResult = _app.Marked("TextResult");
+			Assert.IsNotNull(theTextResult.GetDependencyPropertyValue("Text"));
```
[the remark this answers is not in the document]
Comment (GuillaumeSE) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ListViewTests/UnoSamples_Tests.ListView.cs]: I have fixed it with screenshot instead where a success text is expected

**Pull request merged**